### DISCUSSION TITLE
Unexpected visibility check

### DISCUSSION THREAD

**Discussion by nnaydenow** [cypress-io/cypress#30475] [Questions and Help]
Hi,

I'm working on testing a piece of code, which you can view here: [JSBin Example](https://jsbin.com/sivakuvuja/edit?html,output). The example includes web components and composition, where I aim to reuse a `popover-component` and display a `menu-component` with items. I've also implemented the same setup using native elements.

When running tests with Cypress, I'm encountering unexpected behavior with the `"be.visible"` assertion on these elements.

Here's the code for testing the web components:

```javascript
it('webcomponent', () => {
  cy.visit('https://output.jsbin.com/sivakuvuja');

  cy.get("#btn")
    .click();

  cy.get("menu-component")
    .should("be.visible");

  cy.get("menu-component")
    .shadow()
    .find("popover-component")
    .should("be.visible");
});
```

In this case, Cypress fails with an assertion error indicating that `popover-component` is not visible, even though the visibility assertion for `menu-component` passes.

For comparison, here is the test code for the native elements:

```javascript
it('native', () => {
  cy.visit('https://output.jsbin.com/sivakuvuja');

  cy.get("#btn2")
    .click();

  cy.get(".menu-component")
    .should("be.visible");

  cy.get(".menu-component")
    .find(".popover-component")
    .should("be.visible");
});
```

In this test, Cypress fails with an assertion error stating that `.menu-component` is not visible.

One difference between these tests is the element dimensions. In the first test, `menu-component` has dimensions of `0 x 18` (width x height), whereas in the second test, `.menu-component` has dimensions of `1000 x 0`.

I'm confused as to why the visibility assertion passes for the first test, as per [Cypress Visibility Guidelines](https://docs.cypress.io/app/core-concepts/interacting-with-elements#Visibility). Cypress documentation specifies that an element is considered hidden if it has a width or height of 0.

Am I missing something here?

**Comment by portyu9** [comment]
Hi @nnaydenow 

What you observed in 2024 does not match Cypress's current visibility behavior.

As of Cypress 16, Cypress uses the browser's native visibility check plus an explicit bounding-rectangle guard, and an element is considered hidden if **either its width or its height is `0`**. So both a `0 x 18` element and a `1000 x 0` element should currently fail `be.visible`.

If this still reproduces after upgrading, inspect the dimensions of the exact element being asserted after the menu opens:

```js
cy.get('menu-component').then(($el) => {
  console.log($el[0].getBoundingClientRect())
})
```

With web components, the custom-element host and the element inside its shadow root can have different layout boxes, so assert visibility on the element that actually owns the rendered box.

If you are staying on the older Cypress version from this thread, this is best treated as a legacy visibility edge case rather than relying on the old result.

Please mark this answer as accepted if it helped, thank you.
